### PR TITLE
Add fallibility to `Clock` methods

### LINKED ISSUE #25: Fallibility of `Clock` methods for I2C RTC device driver implementations
First of all, thank you for taking the time to develop embedded-time!

I had a quick look at if implementing `Clock` for I2C RTC devices since I [wrote](https://github.com/eldruin/ds1307-rs) [a few](https://github.com/eldruin/ds323x-rs) [drivers](https://github.com/eldruin/mcp794xx-rs) but at the moment this seems not possible due to fallibility of the methods.

As any interaction over I2C can fail for several reasons (there is undergoing discussion about this at https://github.com/rust-embedded/embedded-hal/issues/229), the methods must provide a way to return an error.
This includes `Clock::now()` and by extension also `Clock::delay()`.

For embedded-time I guess an associated `Error` type would make sense so that any implementation can have its own error definitions.

However, this would make sense if `embedded_time::Clock` implementations on top or I2C RTC device drivers are desired and _may_ be unnecessary for HAL (timer/counter-based) implementations. This would be a project focus question.

### PR DESCRIPTION
**Pull request by PTaylor-us**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -102,6 +103,10 @@ pub mod units {
     pub use crate::frequency::units::*;
 }
 
+/// Time-related error
+#[derive(Debug, Eq, PartialEq)]
+pub struct Error;
```
Comment (eldruin) [inline src/lib.rs:108]: Why is the error type a `struct` and not an `enum` as usual where variants can be added?

Comment (PTaylor-us) [reply src/lib.rs:108]: I started with an enum, but basically just couldn't think of any variants to add at this point. There is currently only one failure mode, not being able to read the current instant. Any suggestions you have, would be much appreciated. Perhaps it's worth adding that as a variant is a good starting point? And adding the non_exhaustive attrib?

Comment (PTaylor-us) [reply src/lib.rs:108]: I think I was not thinking correctly about the error types. I need to provide variants that the implementer will use to attempt to describe the reason the call to `now()` failed.

**Comment by PTaylor-us** [comment]
@eldruin You'll find the beginning of my latest attempt at an error handling scheme in this last commit. I would love to hear if you think I'm at least on the right track since our discussion.

**Comment by eldruin** [comment]
Sounds good. However, I think the `Error` trait is unnecessary. That would be necessary for trait-method-based error handling. i.e. `err.is_other()` but that would be a different way from what we discussed.
Here is what I think should suffice:
```rust
#[non_exhaustive]
#[derive(Debug, Eq, PartialEq)]
pub enum Error<E> {
    /// specific implementation error
    Other(E),
}

pub trait Clock {
    type ImplError: core::fmt::Debug;

    fn now(&self) -> Result<u32, Error<Self::ImplError>>;
}

#[derive(Debug, Eq, PartialEq)]
pub enum BadClockImplError {
    NotStarted,
    OtherError,
}

pub struct BadClock;
impl Clock for BadClock {
    type ImplError = BadClockImplError;

    fn now(&self) -> Result<u32, Error<Self::ImplError>> {
        Err(Error::Other(BadClockImplError::NotStarted))
    }
}

fn main() {
    let clk = BadClock {};
    match clk.now() {
        Err(Error::Other(BadClockImplError::NotStarted)) => println!("Not started!"),
        Err(Error::Other(_)) => println!("Something else"),
        Ok(v) => println!("All good! {}", v),
    }
}
```
This example in the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=797af07382f579fa0fd2f8a9ee65e064).
This should allow you to remove the `Error` trait, related impls and so on. Let me know if I should send you a PR doing this. Maybe that is clearer.

**Comment by PTaylor-us** [comment]
> I think the Error trait is unnecessary.

I agree that it's marginally useful. I added it to collect any traits (currently only fmt::Debug) that would be necessary. I also don't think it would be possible to use `()` as an `ImplError` without it. I realize that `()` probably isn't optimal to use, but it's very helpful to use in doc examples and tests. Having the `Error` trait also leaves open the option of adding not just other supertraits, but also the trait-method-based error handling you mentioned.

> this only makes sense if you plan to add other variants to `Error`

I want to keep that option open to add other variants.

As always, thank you for your guidance on these important considerations.

**Comment by eldruin** [comment]
I see. Ok. You can use `core::convert::Infallible` instead of `()`, which should better convey the meaning.

For the record, I think the trait-method-based approach would not make much sense in its current form: If you require both the `Clock::Error` and `ImplError`  to implement the `Error` trait, both would need to implement something like `is_other()` but that is only a property that makes sense to ask the `Clock::Error`. An `ImplError` is by definition an "other" case since it is contained in the `Clock::Error::Other(E)`. This may be clearer if you think of some other method like `is_overflow()`.
Anyway that would be something for the future.

**Comment by PTaylor-us** [comment]
> You can use core::convert::Infallible instead of (), which should better convey the meaning.

Just this morning I realized I could use `Infallible`, however that also requires impling `Error` for it.

> If you require both the Clock::Error and ImplError to implement the Error trait, both would need to implement something like is_other() but that is only a property that makes sense to ask the Clock::Error. An ImplError is by definition an "other" case since it is contained in the Clock::Error::Other(E). 

Great point. I can't think of any trait methods that might make sense other than, perhaps something meta-level similar to the `std::error::Error` trait methods.

**Pull request merged**